In [59]:
import pandas as pd
import numpy as np

file_path = "../Data/raw/Sample Data for Skills Assessment June 2025.xlsx"

data = pd.read_excel(file_path, sheet_name= "Dataset Client Insights Analyst") 


In [60]:
data.shape



(2015, 15)

In [61]:
data.head()




,Client Key,Client Account ID,Client Name,Age Range,Date of Birth,Gender,Education Level,status,Client Type,Date Opened,Language Primary,Parish,Email Address,Subsidiary Description,Total Funds Under Management
0,10012,2020,Billie(Deceased),25-34,1987-01-28 00:00:00,Male,High school diploma/GED,Inactive,Individual,2017-01-28,English,St. Catherine,Billie@yahoo.com,JMMB Investments (Jamaica),2294.85
1,10013,2022,Charlie,25-34,1986-02-26 00:00:00,Male,Trade/Technical,Inactive,Individual,2017-01-27,English,Kingston,Charlie@yahoo.com,JMMB Investments (Jamaica),959754.98
2,10014,2023,Cleo,25-34,1987-05-31 00:00:00,Female,Some college education,Inactive,Individual,2017-01-28,English,St. Annrew,Cleo@yahoo.com,JMMB Investments (Jamaica),1099.45
3,10015,1002,Olivia,25-34,1992-02-19 00:00:00,Female,Some college education,Inactive,Individual,2017-01-28,English,Kingston,Olivia@yahoo.com,JMMB Investments (Jamaica),227000.00
4,10015,1006,Olivia,25-34,1992-02-19 00:00:00,Female,Some college education,Inactive,Individual,2017-01-28,English,Kingston,Olivia@yahoo.com,JMMB Investments (Jamaica),227000.00


In [62]:

data.tail()

,Client Key,Client Account ID,Client Name,Age Range,Date of Birth,Gender,Education Level,status,Client Type,Date Opened,Language Primary,Parish,Email Address,Subsidiary Description,Total Funds Under Management
2010,10906,6649,Maksymilian,25-34,1989-08-30 00:00:00,Male,No high school,Active,Individual,2015-11-25,English,St. Ann,Maksymilian@gmail.com,JMMB Bank,199202.21
2011,10907,6650,Marc,25-34,1990-12-12 00:00:00,Male,College degree,Active,Individual,2000-06-19,Other,St. Andrew,Marc@gmail.com,JMMB Bank,23418.02
2012,10908,6653,Maximus,25-34,1986-08-04 00:00:00,Male,Some college education,Active,Individual,2015-04-25,English,Florida,Maximus@gmail.com,JMMB Bank,13868.65
2013,10908,6656,Maximus,25-34,1986-08-04 00:00:00,Male,Some college education,Active,Individual,2015-04-25,English,Florida,Maximus@gmail.com,JMMB Bank,13868.65
2014,10909,6681,Niven,25-34,1985-01-28 00:00:00,Male,Some college education,Active,Individual,2017-05-06,English,St. Andrew,Niven@gmail.com,JMMB Bank,1140.05


In [63]:

data.columns.tolist()




['Client Key',
 'Client Account ID',
 'Client Name',
 'Age Range',
 'Date of Birth',
 'Gender',
 'Education Level',
 'status',
 'Client Type',
 'Date Opened',
 'Language Primary',
 'Parish',
 'Email Address',
 'Subsidiary Description',
 'Total Funds Under Management ']

In [64]:
data.info()




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2015 entries, 0 to 2014
Data columns (total 15 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   Client Key                     2015 non-null   int64         
 1   Client Account ID              2015 non-null   int64         
 2   Client Name                    2015 non-null   object        
 3   Age Range                      2015 non-null   object        
 4   Date of Birth                  2015 non-null   object        
 5   Gender                         2015 non-null   object        
 6   Education Level                1988 non-null   object        
 7   status                         2015 non-null   object        
 8   Client Type                    2015 non-null   object        
 9   Date Opened                    2015 non-null   datetime64[ns]
 10  Language Primary               2012 non-null   object        
 11  Parish           

In [65]:
data.isnull().sum()



Client Key                        0
Client Account ID                 0
Client Name                       0
Age Range                         0
Date of Birth                     0
Gender                            0
Education Level                  27
status                            0
Client Type                       0
Date Opened                       0
Language Primary                  3
Parish                            0
Email Address                     0
Subsidiary Description            0
Total Funds Under Management      0
dtype: int64

In [66]:
data.duplicated().sum()



0

In [67]:
data["Client Key"].nunique()



898

WE really only have 898 clients 

client key != rows 

In [68]:
data["Client Key"].duplicated().sum()

1117

In [69]:
data["Client Account ID"].nunique()

2015

Delete redundant columns ,
Dropping duplicates ,
Cleaning Columns, 
Checking and validating emails ,
There are No exact duplicate rows, multiple records share the same Client Key. Maybe This means that clients may have multiple accounts ( IT DOES SHOW THAT EACH ACCOUNT ID is UNIQUE!!!)

INVESTIGATING EACH CATEGORICAL VALUE 
noticed misspelling in client key 2 as St.Annrew

In [70]:
data["Gender"].value_counts(dropna=False)

Gender
Female            987
Male              936
Not Applicable     92
Name: count, dtype: int64

In [71]:
data["status"].value_counts(dropna=False)

status
Inactive    1119
Active       896
Name: count, dtype: int64

In [72]:
data["Parish"].value_counts(dropna=False)

Parish
St. Catherine               421
St. James                   179
Kingston                    151
St. Andrew                  149
Manchester                  138
                           ... 
Kingston 9                    1
Turks and Caicos Islands      1
Mississippi                   1
Ohio                          1
Grand Cayman                  1
Name: count, Length: 64, dtype: int64

In [73]:
data["Age Range"].value_counts(dropna=False)

Age Range
25-34             1355
55-65              381
35-44              112
Not Applicable      92
45-54               75
Name: count, dtype: int64

In [74]:
data["Subsidiary Description"].value_counts(dropna=False)

Subsidiary Description
JMMB Investments (Jamaica)    1096
JMMB Bank                      919
Name: count, dtype: int64

Im trying to see how many clients (client_keys) Have different Client Account IDs  before I drop any duplicates 

In [75]:
client_summary = (
    data.groupby("Client Key").agg(
        account_count = ("Client Account ID", "nunique"),
        subsidiary_count = ("Subsidiary Description", "nunique"),
    )
)
client_summary.describe()

,account_count,subsidiary_count
count,898.000000,898.0
mean,2.243875,1.0
std,1.510126,0.0
min,1.000000,1.0
25%,1.000000,1.0
50%,2.000000,1.0
75%,3.000000,1.0
max,9.000000,1.0


In [76]:
client_summary.sort_values("account_count", ascending=False).head(10)

,account_count,subsidiary_count
Client Key,,
10455,9,1
10252,9,1
10036,6,1
10437,6,1
10037,6,1
10367,6,1
10039,6,1
10040,6,1
10300,6,1


ALRIGHT TIME FOR MISSING VALs ()

In [77]:
data[data["Education Level"].isnull()]["Client Type"].value_counts()


Client Type
Individual    27
Name: count, dtype: int64

In [78]:
data[data["Language Primary"].isnull()]["Client Type"].value_counts()

Client Type
Individual    3
Name: count, dtype: int64

Cleaning Column Names 

In [79]:
cleaned = data.copy()

In [80]:
cleaned.columns =(
    cleaned.columns
    .str.strip()
)
print(cleaned.columns.tolist())

['Client Key', 'Client Account ID', 'Client Name', 'Age Range', 'Date of Birth', 'Gender', 'Education Level', 'status', 'Client Type', 'Date Opened', 'Language Primary', 'Parish', 'Email Address', 'Subsidiary Description', 'Total Funds Under Management']


In [81]:
text_columns = cleaned.select_dtypes(include="object").columns

cols_to_strip = [c for c in text_columns if c not in ["Date of Birth", "Date Opened"]]

for column in cols_to_strip:
    cleaned[column] = cleaned[column].str.strip()

In [82]:
cleaned["Parish"] = cleaned["Parish"].str.title()
cleaned["Language Primary"] = cleaned["Language Primary"].str.title()

Apparently converting to DateTye 64 id better than keeping it as a object 

In [83]:
cleaned["Date of Birth"] = pd.to_datetime(
    cleaned["Date of Birth"], 
    dayfirst=True, 
    errors="coerce"
)

In [84]:
cleaned["Date Opened"] = pd.to_datetime(
    cleaned["Date Opened"], 
    dayfirst=True, 
    errors="coerce"
)

In [85]:
cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2015 entries, 0 to 2014
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   Client Key                    2015 non-null   int64         
 1   Client Account ID             2015 non-null   int64         
 2   Client Name                   2015 non-null   object        
 3   Age Range                     2015 non-null   object        
 4   Date of Birth                 1923 non-null   datetime64[ns]
 5   Gender                        2015 non-null   object        
 6   Education Level               1988 non-null   object        
 7   status                        2015 non-null   object        
 8   Client Type                   2015 non-null   object        
 9   Date Opened                   2015 non-null   datetime64[ns]
 10  Language Primary              2012 non-null   object        
 11  Parish                        

MISSING VALUES ... YAY!

In [86]:
cleaned["Education Level"] = cleaned["Education Level"].fillna("Unknown")
cleaned["Language Primary"] = cleaned["Language Primary"].fillna("Unknown")

In [87]:
cleaned.isnull().sum()

Client Key                       0
Client Account ID                0
Client Name                      0
Age Range                        0
Date of Birth                   92
Gender                           0
Education Level                  0
status                           0
Client Type                      0
Date Opened                      0
Language Primary                 0
Parish                           0
Email Address                    0
Subsidiary Description           0
Total Funds Under Management     0
dtype: int64

ASKED COPILOT HOW I WOULD GO ABOUT STANDERDIZING THE PARISH COLUMN AND IT SUGGESTED THE FOLLOWING:

In [88]:
sorted(cleaned["Parish"].unique())

['Alabama',
 'Arizona',
 'Birmingham',
 'Bridgeport P O',
 'Clarendon',
 'Claxton',
 'England',
 'Florida',
 'Georgia',
 'Grand Cayman',
 'Gregory Park P.',
 'Hanover',
 'Illinois',
 'Kingston',
 'Kingston 10',
 'Kingston 11',
 'Kingston 13',
 'Kingston 14',
 'Kingston 17',
 'Kingston 19',
 'Kingston 2',
 'Kingston 20',
 'Kingston 3',
 'Kingston 4',
 'Kingston 5',
 'Kingston 6',
 'Kingston 7',
 'Kingston 8',
 'Kingston 9',
 'Manchester',
 'Massachusetts',
 'Mississippi',
 'New York',
 'Ohio',
 'Ontario',
 'Pennsylvania',
 'Port Antonio',
 'Port-Of-Spain',
 'Portland',
 'Spanish Town',
 'Spanish Town P.',
 'St  James',
 'St Ann',
 'St Catherine',
 'St.  Ann',
 'St. Andrew',
 'St. Ann',
 'St. Annrew',
 'St. Catherine',
 'St. Elizabeth',
 'St. James',
 'St. Mary',
 'St. Thomas',
 'Surrey',
 'Texas',
 'Trelawny',
 'Turks And Caicos Islands',
 'Westmoreland']

In [89]:
# Standardize parish names
cleaned["Parish"] = (
    cleaned["Parish"]
    .astype("string")
    .str.strip()
    .str.lower()
)

parish_mapping = {
    'kingston 2': 'kingston',
    'kingston 3': 'kingston',
    'kingston 4': 'kingston',
    'kingston 5': 'kingston',
    'kingston 6': 'kingston',
    'kingston 7': 'kingston',
    'kingston 8': 'kingston',
    'kingston 9': 'kingston',
    'kingston 10': 'kingston',
    'kingston 11': 'kingston',
    'kingston 12': 'kingston',
    'kingston 13': 'kingston',
    'kingston 14': 'kingston',
    'kingston 15': 'kingston',
    'kingston 16': 'kingston',
    'kingston 17': 'kingston',
    'kingston 18': 'kingston',
    'kingston 19': 'kingston',
    'kingston 20': 'kingston',
    'bridgeport p o': 'st. catherine',
    'gregory park p.': 'st. catherine',
    'st. annrew': 'st. andrew',
    'spanish town':'st. catherine',
    'spanish town p.': 'st. catherine',
    'st catherine': 'st. catherine',
    'st. andrew': 'st. andrew',
    'st. catherine': 'st. catherine'
    }

cleaned["Parish"] = cleaned["Parish"].replace(parish_mapping)

    


In [90]:
sorted(cleaned["Parish"].dropna().unique())

['alabama',
 'arizona',
 'birmingham',
 'clarendon',
 'claxton',
 'england',
 'florida',
 'georgia',
 'grand cayman',
 'hanover',
 'illinois',
 'kingston',
 'manchester',
 'massachusetts',
 'mississippi',
 'new york',
 'ohio',
 'ontario',
 'pennsylvania',
 'port antonio',
 'port-of-spain',
 'portland',
 'st  james',
 'st ann',
 'st.  ann',
 'st. andrew',
 'st. ann',
 'st. catherine',
 'st. elizabeth',
 'st. james',
 'st. mary',
 'st. thomas',
 'surrey',
 'texas',
 'trelawny',
 'turks and caicos islands',
 'westmoreland']

In [91]:
cleaned["Parish"].value_counts()

Parish
kingston                    540
st. catherine               439
st. james                   179
st. andrew                  150
manchester                  143
clarendon                   121
st. elizabeth                73
st. ann                      69
st. mary                     51
florida                      38
trelawny                     36
new york                     25
westmoreland                 19
portland                     18
st. thomas                   15
ontario                      14
hanover                      13
georgia                       9
england                       6
port antonio                  6
st.  ann                      6
alabama                       6
st ann                        5
st  james                     5
illinois                      5
arizona                       5
texas                         3
surrey                        3
port-of-spain                 3
birmingham                    2
claxton                       2
t

In [92]:
sorted(cleaned["Parish"].unique())

['alabama',
 'arizona',
 'birmingham',
 'clarendon',
 'claxton',
 'england',
 'florida',
 'georgia',
 'grand cayman',
 'hanover',
 'illinois',
 'kingston',
 'manchester',
 'massachusetts',
 'mississippi',
 'new york',
 'ohio',
 'ontario',
 'pennsylvania',
 'port antonio',
 'port-of-spain',
 'portland',
 'st  james',
 'st ann',
 'st.  ann',
 'st. andrew',
 'st. ann',
 'st. catherine',
 'st. elizabeth',
 'st. james',
 'st. mary',
 'st. thomas',
 'surrey',
 'texas',
 'trelawny',
 'turks and caicos islands',
 'westmoreland']

What even is a valid email?? .... Lets find out / how does that look well i guess you kinda know / BUT LETS ASK COLAB TO CREATE A email validation rule based on the article i found this was validated by checking the amount of emails that were true 

In [93]:
import re

def validate_email(email):
    """
    Validate email format using regex pattern.
    Returns True if valid, False otherwise.
    """
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return re.match(pattern, email) is not None

cleaned["validate_email"] = cleaned["Email Address"].apply(validate_email)


Contactability

In [94]:
import re

# ── CONTACTABILITY COLUMN ──
# Add this AFTER validate_email is already in your dataframe

def is_contactable(row):
    """
    Returns True if client can be reached. False if deceased, placeholder email, or invalid format.
    """
    name  = str(row.get("Client Name", "")).lower()
    email = str(row.get("Email Address", "")).lower()
    email_is_valid = row.get("validate_email", False)
    
    # 1. Deceased
    if re.search(r'\bdeceased\b|\(deceased\)|deceased[_-]|[_-]deceased', name):
        return False
    
    # 2. Placeholder emails (local part only)
    local = email.split('@')[0] if '@' in email else email
    placeholders = [
        r'^unknown\d*$', r'^unkown\d*$', r'^unknwn\d*$',
        r'^placeholder\d*$', r'^na\d*$', r'^n/a\d*$',
        r'^none\d*$', r'^test\d*$', r'^noemail\d*$',
        r'^temp\d*$', r'^fake\d*$', r'^dummy\d*$',
        r'^notprovided\d*$', r'^not_provided\d*$'
    ]
    if any(re.search(p, local) for p in placeholders):
        return False
    
    # 3. Invalid format
    if not email_is_valid:
        return False
    
    return True

cleaned["Contactable"] = cleaned.apply(is_contactable, axis=1)

# ── QUICK CHECK ──
print(cleaned[["Client Key", "Client Name", "Email Address", "validate_email", "Contactable"]].head(10))

   Client Key       Client Name      Email Address  validate_email  \
0       10012  Billie(Deceased)   Billie@yahoo.com            True   
1       10013           Charlie  Charlie@yahoo.com            True   
2       10014              Cleo     Cleo@yahoo.com            True   
3       10015            Olivia   Olivia@yahoo.com            True   
4       10015            Olivia   Olivia@yahoo.com            True   
5       10015            Olivia   Olivia@yahoo.com            True   
6       10015            Olivia   Olivia@yahoo.com            True   
7       10015            Olivia   Olivia@yahoo.com            True   
8       10015            Olivia   Olivia@yahoo.com            True   
9       10016   Emily(Deceased)    Emily@yahoo.com            True   

   Contactable  
0        False  
1         True  
2         True  
3         True  
4         True  
5         True  
6         True  
7         True  
8         True  
9        False  


In [95]:
cleaned["Date of Birth"] = pd.to_datetime(
    cleaned["Date of Birth"], 
    dayfirst=True, 
    errors="coerce"
)

In [96]:
cleaned.shape

(2015, 17)

In [97]:
cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2015 entries, 0 to 2014
Data columns (total 17 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   Client Key                    2015 non-null   int64         
 1   Client Account ID             2015 non-null   int64         
 2   Client Name                   2015 non-null   object        
 3   Age Range                     2015 non-null   object        
 4   Date of Birth                 1923 non-null   datetime64[ns]
 5   Gender                        2015 non-null   object        
 6   Education Level               2015 non-null   object        
 7   status                        2015 non-null   object        
 8   Client Type                   2015 non-null   object        
 9   Date Opened                   2015 non-null   datetime64[ns]
 10  Language Primary              2015 non-null   object        
 11  Parish                        

In [98]:
cleaned.isnull().sum()

Client Key                       0
Client Account ID                0
Client Name                      0
Age Range                        0
Date of Birth                   92
Gender                           0
Education Level                  0
status                           0
Client Type                      0
Date Opened                      0
Language Primary                 0
Parish                           0
Email Address                    0
Subsidiary Description           0
Total Funds Under Management     0
validate_email                   0
Contactable                      0
dtype: int64

In [99]:
cleaned.to_excel("../Data/processed/cleaned_client_data_Final.xlsx", index=False)